[PC-1197] Wordpress and Database Portenta X8 Tutorial Update (#949)
arduino/docs-content: content/hardware/04.pro/boards/portenta-x8/tutorials/13.wordpress-webserver/content.md

- [Arduino Portenta X8](https://store.arduino.cc/products/portenta-x8)
- USB-C® cable (either USB-C® to USB-A or USB-C® to USB-C®)
- The [docker-compose.yml](assets/docker-compose.rar) file used in this tutorial

## Instructions

First, make sure your Portenta X8 is set up correctly by following the [User Manual's Out-of-the-Box Experience](https://docs.arduino.cc/tutorials/portenta-x8/user-manual#out-of-the-box-experience).

### Creating the **docker-compose.yml** File

> The WordPress container we use is a multi-container application, which also requires a database server container. The WordPress multi-container application uses Apache as its web server. This is required to make the service work and it is already included in the container, so it is nothing for us to worry about. We will be using **MariaDB** as our database server container.

Reviewer: ```suggestion
The WordPress container we use is a multi-container application, which also requires a database server container. The WordPress multi-container application uses Apache as its web server. It is required to make the service operational and comes included within the container. We will be using MariaDB as our database server as a container instance.
```
Suggested section rephrase.

> This container can run on the Portenta X8's architecture. All we need to start using these containers is to write a **docker-compose.yml** file. This file will contain information about what image we want to install and some important configuration information, such as the username for the database, password, timezone and database name. The same goes for the WordPress container: it will contain the password and username and we will also enter the database hostname and which container will be used as the database. We recommend that you change the default passwords to more secure ones by replacing the default ones that are stated in the file below.

Reviewer: ```suggestion
This container can run on the Portenta X8's architecture. To start using these containers is to write a docker-compose.yml file. This file contains information regarding what image we want to install and important configuration details, such as the username for the database, password, timezone, and database name. The same goes for the WordPress container: it will include the password and username. We will also enter the database hostname and which container will be used as the database. We recommend changing the default passwords to more secure ones by replacing the default password defined in the file below.
```

Suggested section rephrase.

      WORDPRESS_DB_PASSWORD: Wordpress
      WORDPRESS_DB_NAME: Wordpress
volumes:
  Wordpress_data: {}
  db_data: {}
```

Now let's create a directory on our X8 and put this **docker-compose.yml** file on our device. You can download the file by clicking [here](assets/docker-compose.rar).

### Installing The Containers

First, we create a directory where we want to add our **docker-compose.yml** file. Using the `mkdir` command we will create a directory named "wordpress-test". Navigate into this directory with a simple `cd` command. Either copy the docker-compose.yml file into this directory or create it directly here.

To create the file, we can use `cat > docker-compose.yml`, this will create the file, so you can copy the content of the file from above and paste it. Push enter once to go to a new line and press `ctrl C` to exit the file editor. To copy the file from your computer onto the device use:

```
adb push <path to docker-compose.yml file> /home/fio/wordpress-test
```

Alternatively, you could place the `docker-compose.yml` file inside the `wordpress-test` directory and push the file using the following command:

```
adb push .\wordpress-test\ /home/fio
```

> Both options work fine and depend on how you would like to handle the file.

Reviewer: I have separated the paragraph a bit for better readability since it has a lot of commands that users need to follow, especially the main commands. Also added the last information fragment for an additional alternative which was verified while testing the code. How do you see this change? Please feel free to revert the change if you would like to.

![cd into correct directory](assets/webserver-mkdir.png)

***Remember that you may need to run the next command to gain admin access for running the Docker's commands: `sudo su -`  which default password is `fio`***

Before installing the containers, make sure that no other container is running on the ports that the WordPress container will use. You can check what containers are running and what port they are using by running the `docker ps -a` command. This will show a list of the currently installed and running containers on the Portenta X8.

> To remove a container first stop it with `docker stop <container id>`, then you can run `docker rm <container id>` to remove it. If you want more information about handling containers on your Portenta X8, take a look at our [Managing Containers with Docker tutorial](https://docs.arduino.cc/tutorials/portenta-x8/docker-container).

Reviewer: I have separated the paragraph a bit for better readability since it has a lot of commands that users need to follow. How do you see this change? Please feel free to revert the change if you would like to.

When you are in the correct directory and no other container is running on the ports that WordPress will use, you can now run `docker compose up -d`. Using the `-d` tag in the command will allow running these containers in the background. If you run the command without the `-d` tag, the application will exit when you close the terminal.

When the command is executed it will start installing the **WordPress** and **MariaDB** containers. This can take a while. To get the output from the containers use: `docker-compose logs -f`. Once it is done you can connect to the device and site.

![Containers install progress in the terminal](assets/webserver-container-install.png)

### Connecting to the WordPress Site

> To connect to the WordPress setup site, you simply need to access it with your Portenta X8s unique id and port. So we can use the following address format:

Reviewer: ```suggestion
To connect to the WordPress setup site, you simply need to access it with your Portenta X8's unique id and port. So we can use the following address format:
```
Added missing apostrophe.

```

> http://portenta-x8-<uuid>.local:<port>

Reviewer: Here, I think it could be nice to show the user that they can obtain the uuid by looking at the root id. For example, using the current tutorial's example image, the uuid is `3321ba09dab6fad9`. So the address is now "http://portenta-x8-3321ba09dab6fad9.local:8000". What do you think about this suggestion?

> http://portenta-x8-<uuid>.local:<port>

Reviewer: Here, I think it could be nice to show the user that they can obtain the uuid by looking at the root id. For example, using the current tutorial's example image, the uuid is `3321ba09dab6fad9`. So the address is now "http://portenta-x8-3321ba09dab6fad9.local:8000". What do you think about this suggestion?

Author: I didn't know that! It is really interesting. Thanks!

> To connect to the WordPress setup site, you simply need to access it with your Portenta X8s unique id and port. So we can use the following address format:
> 
> ```
> http://portenta-x8-<uuid>.local:<port>
> ```

Reviewer: ```suggestion
To connect to the WordPress setup site, you simply need to access it with your Portenta X8's unique id and port. So we can use the following address format:

```
http://portenta-x8-<uuid>.local:<port>
```
```

Here, I think it could be nice to show the user that they can obtain the uuid by looking at the root id. For example, using the current tutorial's example image, the uuid is `3321ba09dab6fad9`. So the address is now "http://portenta-x8-3321ba09dab6fad9.local:8000". What do you think about this suggestion? Also included a minor fix.

> Where you would substitute the `<uuid>` with your Portenta X8's unique id and the port chosen for the WordPress container with `<port>`. The `<uuid>` can be found on the setup page that is shown in the [User Manual's Out-of-the-Box Experience](https://docs.arduino.cc/tutorials/portenta-x8/user-manual#out-of-the-box-experience), but you can also get it in the terminal when running `adb` or you can go to `http://192.168.7.1:8000` if you use Windows and Linux, on MacOS use `http://192.168.8.1:8000`.

Reviewer: I have separated the paragraph a bit for better readability since it has a lot of commands that users need to follow, especially the main commands. How do you see this change? Please feel free to revert the change if you would like to.

  Wordpress:
    depends_on:
      - db

>     image: wordpress:latest

Author: @jcarolinares This capital letter was the problem

    volumes:
      - Wordpress_data:/var/www/html
    ports:
      - "8000:80"
    restart: always
    environment:

>       WORDPRESS_DB_HOST: db

Reviewer 2: you sure about this?

Like you are changing the enviromental variables to caps, is this breaking the program or???

Reviewer: I was wondering about this too. And from what I was able to replicate a successful result, it is working. So I think there is no problem.

You are now free to go through the WordPress setup process and configure it however you like.

### Removing the containers ###

If you want to remove the container, you have to go to ```/home/fio/wordpress-test``` directory (where we previously executed the docker-compose command) and execute the following commands according to your needs:

Remove the container but preserves your WordPress database:

```
docker compose down
```

Remove the container and the database:

```
docker compose down --volumes
```

To make sure that it was successful, run ```docker ps -a``` and check that the WordPress and MariaDB containers have disappeared.

## Conclusion

In this tutorial, we went through how to install and run a WordPress and database container on the Portenta X8. We then accessed the WordPress site on our X8 through our web browser. So now you can set up your own WordPress site on your X8 device and access it from another device.

> ## Troubleshooting

Reviewer: In this Troubleshooting section, the third bullet point's command is giving an error in argument and we could use instead `ip -h address`. What do you think about this?

> 

Reviewer: In this Troubleshooting section, the third bullet point's command is giving an argument error and we could use instead ip -h address. What do you think about this?

Reviewer: Also, it does not let me mark the changes for the last line, which is the third bullet point. Please note required changes as follows:

- Apostrophe for "... double-check the X8's IP ..."
- Trademark for Wi-Fi®: "... via USB and Wi-Fi®. ..."

Author: Done and commited. Thank you so much!
